checker: unwrapping a generic option to a specific fallback value should be an error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 602b097.e32e9f7
OS: macos, macOS, 15.1.1, 24B2091
Processor: 10 cpus, 64bit, little endian, Apple M4

getwd: /Users/leodev/p/v/fast2
vexe: /Users/leodev/p/v/v/v
vexe mtime: 2024-12-07 11:48:53

vroot: OK, value: /Users/leodev/p/v/v
VMODULES: OK, value: /Users/leodev/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.47.0
Git vroot status: weekly.2024.49-33-ge32e9f70
.git/config present: true

CC version: Apple clang version 16.0.0 (clang-1600.0.26.4)
emcc version: N/A
thirdparty/tcc status: thirdparty-macos-arm64 713692d4
```

**What did you do?**

I assigned a string to an int.

`./v -g -o vdbg cmd/v && ./vdbg repro.v`

```v
struct Foo[T] {
pub:
	value ?T
}

fn (f Foo[T]) str() string {
	value := f.value or { 'this should be a checker error' }
	return '${value}'
}

fn main() {
	bar := Foo[int]{
		value: 0
	}

	println(bar)
}
```

**What did you expect to see?**

A checker error

**What did you see instead?**

```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/repro.01JEGG3KPG3VD5CZD37E657F9T.tmp.c:5804:20: error: assigning to 'int' from incompatible type 'string' (aka 'struct string')
cc:  5804 |                 *(int*) _t1.data = _SLIT("this should be a checker error");
cc:       |                                  ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc: 1 error generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzU0M2EwMWQyZWY0Y2JjYzQ2ZjdhODQiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.QxtYgIBLOmEv-q-3K84BBcG12PXQJVvnif8WmolI2eo">Huly&reg;: <b>V_0.6-21526</b></a></sub>

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1720,9 +1720,10 @@ fn (mut c Checker) selector_expr(mut node ast.SelectorExpr) ast.Type {
 		}
 		node.typ = field.typ
 		if node.or_block.kind == .block {
-			c.expected_or_type = node.typ.clear_option_and_result()
+			unwrapped_typ := c.unwrap_generic(node.typ)
+			c.expected_or_type = unwrapped_typ.clear_option_and_result()
 			c.stmts_ending_with_expression(mut node.or_block.stmts, c.expected_or_type)
-			c.check_or_expr(node.or_block, node.typ, c.expected_or_type, node)
+			c.check_or_expr(node.or_block, unwrapped_typ, c.expected_or_type, node)
 			c.expected_or_type = ast.void_type
 		}
 		return field.typ

--- a/vlib/v/checker/tests/selector_generic_or_block_err.out
+++ b/vlib/v/checker/tests/selector_generic_or_block_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/selector_generic_or_block_err.vv:7:24: error: wrong return type `string` in the `or {}` block, expected `int`
+    5 | 
+    6 | fn (f Foo[T]) str() string {
+    7 |     value := f.value or { 'this should be a checker error' }
+      |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    8 |     return '${value}'
+    9 | }

--- a/vlib/v/checker/tests/selector_generic_or_block_err.vv
+++ b/vlib/v/checker/tests/selector_generic_or_block_err.vv
@@ -1,0 +1,17 @@
+struct Foo[T] {
+pub:
+	value ?T
+}
+
+fn (f Foo[T]) str() string {
+	value := f.value or { 'this should be a checker error' }
+	return '${value}'
+}
+
+fn main() {
+	bar := Foo[int]{
+		value: 0
+	}
+
+	println(bar)
+}

